### Ejercicios Segundo Parcial

Problemas; 6.1, 6.2, 6.6, 6.7, 6.10, 6.14, 6.22, 6.27, 6.29, 6.30, 6.43, 6.44,
6.47, 6.49, 6.52, 6.53

Considera el siguiente problema ->
<br> 
<br>
Maximiza 
<br>
$-x_1 + 2x_2$
<br>
Sujeto a 

$3x_1 + 4x_2 < 12$
<br>
$2x_1 - x_2 > 2$
$x_1$ , $x_2 > 0 $


<p>Resolver por método gráfico -></p>
<center><img src="./captura_1.jpg" style="width:500px"></center>

In [2]:
from dual_simplex import to_objective_function_value,simplex,dual_simplex_method
c = [-1, 2, 0, 0]
A = [
    [3, 4, 1, 0],
    [-2, 1, 0, 1]
    ]
b = [12, -2]

primal = to_objective_function_value(c, simplex(c, A, b))
print('Primal: ', primal)

Primal:  1.4545454545454544


In [27]:
primer_resultado = simplex(c,A,b)
primer_resultado

[1.8181818181818181, 1.6363636363636362, 0, 0]

In [30]:
ca = [12,-2,0,0]
Aa = [
    [3,-2, 1, 0],
    [4,1, 0, 1]
]
ba = [-1,2] 

dualb = to_objective_function_value(c, dual_simplex_method(ca, Aa, ba))
print('Dual: ', dualb)

Dual:  1.0


In [31]:
dual_simplex_method(ca,Aa,ba)

[0, 0.5, 0, 1.5]

Como tal en este ejercicio no fue posible pasarlo a primal dual => debido a la restriccion de 2

<p>Ejemplo de resolver primal dual -></p>
<center><img src="./captura_2.jpg" style="width:500px"></center>

Consideral el siguiente problema
<br>

Minimizar
<br>
$2x_1 + 3x_2 + 5x_3 + 6x_4$
Sujeto a
<br> 
$x_1+2x_2+3x_3+x_4>2$
<br>
$-2x_1+2x_2-x_3+3x_4<-3$
<br>
$x_1,x_2,x_3,x_4>0$

In [33]:
c2 = [2,3,5,6, 0, 0]
A2 = [
    [1,2,3,4, 1, 0],
    [1,-1,1,-3, 0, 1]
    ]
b2 = [2, 3]

primal2 = to_objective_function_value(c2, simplex(c2, A2, b2))
print('Primal: ', primal2)

Primal:  4.0


In [34]:
simplex(c2, A2, b2)

[2.0, 0, 0, 0, 0, 1.0]

In [35]:
ca = [-2-3,0,0,0,0]
Aa = [
    [-1-2,1,0,0,0],
    [-2,1,0,1,0,0],
    [-3-1,0,0,1,0],
    [-1,3,0,0,0,1]
]
ba = [2,3,5,6] 

dualb = to_objective_function_value(c, dual_simplex_method(ca, Aa, ba))
print('Dual: ', dualb)

Dual:  0


c:\Users\Adolfo\OneDrive - Instituto Politecnico Nacional\MatematicasAvanzadasCienciaDatos\Ejercicios_Segundo_Parcial\dual_simplex.py:46: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  columns = np.array(tableau).T


No se puede resolver geometricamente , da desbordado => continuamos al siguiente ejercicio

6.6 Considera el siguiente problema

$10x_1 + 24x_2 + 20x_3 +20x_4 +25x_5 $
<br>
$x_1 + x_2 + 2x_3 + 3x_4 + 5x_5 < 19$
<br>
$2x_1 + 4x_2 + 3x_3 + 2x_4 + x_5 < 57$
<br>
$x_1,x_2,x_3,x_4,x_5 > 0$

a) Escribe el problema dual y verifica que ($w_1$ , $w_2$ ) = (4,5) es una solucion feasible.


-------------------------------------

Minimizar 

$19w_1 + 57w_2$
<br>
$w_1 + 2w_2 > 10$
<br>
$w_1 +4w_2 > 24$
<br>
$2w_1 + 3w_2 > 20$
<br>
$3w_1 + 2w_2 > 20$
<br>
$5w_1 + w_2 > 25$
<br>
$w_1,w_2,>0$

4 + 2(5) = 14>10 ---> satisface la restriccion
<br>
4 + 4(5) = 24>24 ---> satisface la restriccion
<br>
2(4) + 3(5) = 23 > 20 ---> satisface la restriccion
<br>
3(4) + 2(5) = 22 > 20 ---> satisface la restriccion
<br>
5(4) + 5 = 25 > 25 
<br>
por lo tanto ( $w_1$ , $w_2$ ) = (4,5) es feasible para el dual

b) Usa la informacion en parte (a) para derivar una solucion optima para ambas , primal y dual 

In [17]:
c3 = [10,24,20,20,25,0,0]
A3 = [
    [1,1,2,3,5,1,0],
    [2,4,3,2,1,0,1]
    ]
b3 = [19, 57]

primal3 = to_objective_function_value(c3, simplex(c3, A3, b3))
print('Primal: ', primal3)

Primal:  361.0


cambiar a negativos las restricciones debido a que cambia a funcion maximizar -> minimizar

In [12]:
ca = [19,57,0,0,0,0,0]
Aa = [
    [-1,-2,1,0,0,0,0],
    [-1,-4,0,1,0,0,0],
    [-2,-3,0,0,1,0,0],
    [-3,-2,0,0,0,1,0],
    [-5,-1,0,0,0,0,1]   
]
ba = [-10,-24,-20,-20,-25]

dualb = to_objective_function_value(ca, dual_simplex_method(ca, Aa, ba))
print('Dual: ', dualb)

Dual:  361.0


6.7 .- Considera el siguiente problema de programacion lineal

P Minimizar $6X_1 + 2X_2$
<br>
Sujeto a:
<br>
$X_1 + 2X_2 > 3$
<br>
$x_2>0$
<br>
$x_1$ sin restricciones 

Para minimizar
<br>
P = $6x_1 + 2x_2$
<br>
Asumimos un plano de 2D  de x1 y x2 to plot linear inequalities and get the common solution.

Sujetos a las restricciones que son:
<br>
$x_1 + 2x_2 > 3$ (1)
<br>
$x_2 > 0$ (2)
<br>
A(3,0)
<br>
B(0,1.5)
<br>
P = 6(3) + 0 
<br>
P = 18
<br>
P = 6(0) + 2(1.5)
<br>
P = 3
<br>
P es minimo en el punto B cuando $x_1 = 0$ , $x_2 = 1.5$
<br>
Dada el problmea lineal para minimizar p sujeto a las constantes son consistentes de dos variables $x_!$ y $x_2$ asi que asumimos un plano 2D con ejes x1 y x2 , dado a ciertos problema.

-------------------


6.10 
Consider the problem: Minimize ex subject to Ax =b, x > 0 where m = n , c = b' and A = A'. Show that if there exists an X0 such that Ax0 = b, then X0 is an optimal point. (Hint. Use duality.)

Sujeto a
<br>
Ax = b, x > 0
<br>
Donde m=n,c = b', A = A'
<br>

Desde la minimizacion necesitamos convertir a la forma estandar para obtener su dual
<br>
maximizar - c´x
<br>
Dado que Ax = b, x > 0
<br>
A si que su dual esta dado
<br>
Minimizado
<br>
Sujeto a A'y = c', y > 0
<br>
Desde que nos dieron c=b', A = A'
<br>
A si que dual se convierte en
<br>
minimizar cy
<br>
Sujeto a Ay = b, y>O
<br>
Clearly constraints of primal and dual are same so we can replace y by x in the dual so dual becomes
<br>
minimize cx
<br>
Subject to Ax = b, x>0
<br>
Now it was given that there exists  such that Axo = b,X>0 clearly  satisfies condition of fesibility so  is feasible solution of the primal.
<br>
Clearly this also satisfies constraints of the dual problem so  is also a feasible solution to the dual.
<br>
Now using theorem of duality
<br>
If x and y are feasible solutions of the primal and dual and cx = by, then x and y must be optimal solutions to the primal and dual
<br>
And we have c=b given
<br>
So CXo = bxo
<br>
Hence must be optimal solution to the primal and dual.

--------------------------

6.14
The Sewel Manufacturing Company produces two types of reclining chairs
for sale in the Northeast. Two basic types of skilled labor are involved
—assembly and finishing. One unit of the top-of-the-line recliner requires
f hours of assembly, 1 hour of finishing, and sells for a profit of $20. A
unit of the second-line recliner requires 1/2 hour of assembly, 1\2 hour of
finishing, and sells for a profit of $12. Presently there are 100 assembly
hours and 80 finishing hours available to the company. The company is
involved in labor negotiations concerning salary modifications for the
coming year. You are asked to provide the company with indications of the
worth of an hour of assembly worker's time and an hour of finishing
worker's time.

Considera el numero de el tope de la linea reclinada sera x y la segunda linea reclinada sera y
<br>
Una unidad de el tope de la line del reclinado toma 2 horas en ensamblar y una hora en terminar. La ganancia en venderlo es de 20
<br>
El numero total de ensambaje es de 100 y 80 horas eso implica que $3/2x + 1y < 100 $ y $ 1/2x + 1/2y < 80 $

La ganancia esta dada por = $f(x,y)$ = $ 20x + 12y $

<p>Ejemplo de resolver primal dual -></p>
<center><img src="./captura_3.jpg" style="width:500px"></center>

20(66.67)+12(0)=1333.33

Observamos que solo necesitamos la primer linea de produccion , la segunda es innecesaria y mas cara
<br>
Utilizando una regla de 3, sabemos que  3/2 = 1333.33 entonces:
<BR>
1 hora de trabajo de ensabmbador esta dada por 888.886666667/100 = 8.88 dolares
<br>
1 hora de acabados de trabajador esta dado por 888.886666667/80 = 11 dolares

------------------


6.22 
Two players are involved in a competitive game. One player, called the row
player, has two strategies available; the other player, called the column
player, has three strategies available. If the row player selects strategy /' and
the column player selects strategy y, the payoff to the row player is c^ and
the payoff to the column player is —
Cy. Thus the column player loses what
the row player wins and vice versa—a two-person zero-sum game. The
following matrix gives the payoffs to the row player.

La neta esta se las debo , esta bien largota :cccccccc 

6.27 Solve the following problem by the dual simplex method.
<br>
Maximize $—4x_1, —6x_2 —18x_3$
<br>
Subject to 
<br>
$x_1 + 3x_3 >3$
<br>
$x_2 + 2x_3 > 4 $
<br>
$x_1, x_2,x_3 > 0$
<br>
Give the optimal values of all the primal and dual variables. Demonstrate
that complementary slackness holds.

--------


Hay que esperar que me la contesten los dudes de la india

6.29 Solve the following linear program by the dual simplex method.

$2x_1 + 3x_2 + 5x_3 + 6x_4$
<br>
Sujeto a 
<br>
$x_1 + 2x_2 + 3x_3 + x_4 > 2$
<br>
$-2x_1 + x_2 -x_3 + 3x_4 > -3$
<br>
$x_1 , x_2 , x_3 , x_4 > 0$

In [30]:
c3 = [2,3,5,6,0,0]
A3 = [
    [1,2,3,1,1,0],
    [2,-1,1,-3,0,1]
    ]
b3 = [2,3]

primal3 = to_objective_function_value(c3, simplex(c3, A3, b3))
print('Primal: ', primal3)

Primal:  12.0


In [22]:
simplex(c3, A3, b3)

[1.8, 0, 0, 0.2, 0, 0]

In [31]:
ca = [2,-3,0,0,0,0]
Aa = [
    [-1,2,1,0,0,0],
    [-2,-1,0,1,0,0],
    [-3,1,0,0,1,0],
    [-1,-3,0,0,0,1]
]
ba = [-2,-3,-5,-6]

dualb = to_objective_function_value(ca, dual_simplex_method(ca, Aa, ba))
print('Dual: ', dualb)

Dual:  12.0


In [32]:
dual_simplex_method(ca, Aa, ba)

[6.0, 0, 4.0, 9.0, 13.0, 0]

6.30 Consider the following problem.
<br>
Minimize 
<br>
$3x_1+5x_2 - x_3 + 2x_4 - 4x_5$
<br>
Sujeto a 
<br>
$x_1 + x_2 + x_3 + 3x_4 + x_5 < 6$
<br>
$-x_1 - x_2 + 2x_3 + x_4 - x_5 > 3$
<br>
$x_1,x_2,x_3,x_4,x_5>0$

In [36]:
c3 = [3,5,-1,2,-4,0,0]
A3 = [
    [1,1,1,3,1,1,0],
    [1,1,-2,-1,1,0,1]
    ]
b3 = [6,-3]

primal3 = to_objective_function_value(c3, simplex(c3, A3, b3))
print('Primal: ', primal3)

Primal:  12.0


In [35]:
simplex(c3, A3, b3)

[0, 5.0, 1.0, 0, 0, 0, 0]

In [45]:
ca = [6,-3,0,0,0,0,0]
Aa = [
     [-1,-1,1,0,0,0,0],
     [-1,-1,0,1,0,0,0],
     [-1,2,0,0,1,0,0],
     [-3,1,0,0,0,1,0],
     [-1,-1,0,0,0,0,1]
     ]
ba = [-3,-5,1,-2,4]

dualb = to_objective_function_value(ca, dual_simplex_method(ca, Aa, ba))
print('Dual: ', dualb)

Dual:  30.0


6.43)

Apply the primal-dual method to the following problem.

Minimize $ 9x_1,-7x_2-4x_3-2x_4-6x_5 - 10x_6$


Sujeto a 

$x_4 + x_5 + x_6 = 8$

$x_1 + x_4 = 6$

$x_2 + x_5 = 4$

$x_3 + x_6 = 3$

$x_1,x_2,x_3,x_4,x_5,x_6 > 0$


In [3]:
c3 = [9,-7,-4,-2,-6,-10,0,0,0,0,0]
A3 = [
    [0,0,0,1,1,1,1,0,0,0,0],
    [0,0,0,1,1,1,0,1,0,0,0],
    [1,0,0,1,0,0,0,0,1,0,0],
    [0,1,0,0,1,0,0,0,0,1,0],
    [0,0,1,0,0,1,0,0,0,0,1]
    ]
b3 = [8,5,6,4,3]

primal3 = to_objective_function_value(c3, simplex(c3, A3, b3))
print('Primal: ', primal3)

Primal:  54.0


In [6]:
ca = [8,5,6,4,3,0,0,0,0,0]

Aa = [
     [-1,0,0,0,0,1,0,0,0,0],
     [-1,0,0,-1,0,0,1,0,0,0],
     [-1,0,0,0,-1,0,0,1,0,0],
     [0,-1,-1,0,0,0,0,0,1,0],
     [0,-1,0,0,-1,-1,0,0,0,1]
     ]

ba = [-9,-7,-4,-2,-6,-10]

dualb = to_objective_function_value(ca, dual_simplex_method(ca, Aa, ba))
print('Dual: ', dualb)

Dual:  94.0


6.44 Resuelve el siguiente problema por el algoritmo primal dual

Minimizar
$x_1 + 2x_3 - x_4$

$x_1 + x_2 + x_3 + x_4 <6$

$2x_1 - x_2 + 3x_3 - 3x_4 > 5$

$x_1,x_2,x_3,x_4 > 0 $


Multiplicamos por menos uno para empezar a trabajarlo

Minimizar


$x_1 + 2x_3 - x_4$

$-x_1 + -x_2 + -x_3 + -x_4 > -6$

$2x_1 - x_2 + 3x_3 - 3x_4 > 5$

$x_1,x_2,x_3,x_4 > 0 $

In [10]:
c3 = [1,0,2,1,0,0]
A3 = [
    [-1,-1,-1,-1,1,0],
    [2,-1,3,-3,0,1]
    ]
b3 = [-6,5]

primal3 = to_objective_function_value(c3, simplex(c3, A3, b3))
print('Primal: ', primal3)

Exception: Linear program is unbounded.

La forma dual esta dada por : 

Maximizar

$-6y_1 + 5y_2$

$-y_1 + 2y_1 < 1$

$-y_1 - y_2 < 0$

$-y_1 + 3y_3 < 2$

$-y_1 - y_2 < -1$

$$


In [7]:
ca = [-6,5,0,0,0,0]

Aa = [
     [-1,2,1,0,0,0],
     [-1,-1,0,1,0,0],
     [-1,-3,0,0,1,0],
     [-1,-1,0,0,0,1]
     ]

ba = [1,0,2,-1]

dualb = to_objective_function_value(ca, dual_simplex_method(ca, Aa, ba))
print('Dual: ', dualb)

Dual:  1.333333333333334


6.47 

<img src="./flojera.jpg" style="width:800px">

Recuerda que todas las >= restricciones pueden ser convertidas a >= multilpicando ambos lados por (-1) 

desde la primera, y tercer restriccion las restricciones $y_1 , y_3$ no seran restringidas en signo

por lo tanto =>

Minimizar zy =  $ 27y_1 -2y_2 + 16y_3$

<img src="./flojera2.jpg" style="width:800px">

In [19]:
ca = [27,-2,16,0,0,0,0,0]

Aa = [
     [-3,-1,-9,1,0,0,0,0],
     [-5,-2,-4,0,1,0,0,0],
     [-6,-3,-2,0,0,1,0,0],
     [-2,-7,-5,0,0,0,1,0],
     [-4,-6,-2,0,0,0,0,1],
     ]

ba = [-7,-2,-1,-4,-6]

dualb = to_objective_function_value(ca, dual_simplex_method(ca, Aa, ba))
print('Dual: ', dualb)

Dual:  41.333333333333336


In [20]:
dual_simplex_method(ca, Aa, ba)

[1.3333333333333333,
 0,
 0.33333333333333337,
 0,
 6.0,
 7.666666666666666,
 0.3333333333333335,
 0]

6.49)

<img src="./tiempo.jpg" style="width:500px">

<img src="./viento.png" style="width:800px">

6.52 Un granjero tiene una superficie de plantio de 500 acres y desea optimizarlos ,  the acreage allocated to the following three crops ; wheat , corn , and soybeans. The man-days , preparation cost y ganancia por acre de los tres plantios  son resumidos a continuacion

| Crop        | Man-Days    |Preparation Cost |   Profit          |
| ----------- | ----------- | -----------    | ----------- |
| Wheat      | 6       |    100            |        60     |
| Corn   | 8        |          150      |          100   |
| SoyBeans  |10  | 120  |80

Suponga que el maximo numero de dias de trabajo de personas dispoinible son 5000 y el granjero tiene $60,000 de preparaciion

a) Encuentre la solucion optima
<br>
b) Asuma que  una jornada de trabajo de 8 horas , va a ser beneficioso al granjero obteniendo una ayuda adicional de $3  por hora ¿porque o porque no?
<br>

| Crop        |Variables| Man-Days    |Preparation Cost |   Profit    |
| ----------- |---------| ----------- | -----------     | ----------- |
| Wheat       |         | 6           | 100             | 60          |
| Corn        |         | 8           | 150             | 100         |
| SoyBeans    |         |10           | 120             | 80          |
| Limit      |         |             | 5000            | 60000       |

Programa Lineal:
<br>
Maximizar $Z$ = $150x_1$ + $80x_2$ + $60x_3$
<br>
Sujeto  a:
<br>
$6x_1 + 8x_2 + 10x_3 <= 5,000$
<br>
$100x_1 + 120x_2 + 100x_3 <= 60,000$
<br>
$x1, x2, x3 >= 0$

Solucion Optima : $x_1 = 200, x_2 = 2,500, x_3 = 2,300$

Tabla optima:

|$Z$   |  150 | 80 | 60  |  0    |
|------|----- |----|-----|-----  |
|$x_1$ |  6   |  8 |  10 |  5,000|
|$x_2$ |  100| 120| 100  |  60,000|
|$x_3$ |  0  | 0 |  0   |  2,300|



b) No , no seria rentable para el granjero obtener ayuda adicional por $3 la hora. La solucion optima indica que el granjero debe colocar todos los recursos disponibles al trigo y al maiz, en orden para maximizar la ganancia

Analisis de Sensibilidad:


Si el granjero ha contratado para entregar al menos lo equivalente de 100 acres de trigo , la nueva solicion optima sera: 

$x_1 = 200, x_2 = 2,500, x_3 = 3,000$

Tabla Optima


|Z  |  150 | 80 | 60  |  0|
|----|-----|-----|----|----|
|x1 |  6 |  8  | 10  |  5,000|
|x2 |  100 |120| 100  |  60,000|
|x3 |  0 |  0  | 0   |  3,000|

The optimal solution indicates that the farmer would need to use more resources to grow wheat, and should allocate additional resources to wheat, instead of corn.
explain
The optimal solution indicates that the farmer would need to use more resources to grow wheat, and should allocate additional resources to wheat, instead of corn.
d.
No, it would not be profitable to the farmer to plant wheat at $5 per bushel. The optimal solution indicates that the farmer should allocate all of the available resources to corn, in order to maximize profit.


La solucion optima indica que el granjero debera necesitar usar más recursos para hacer crecer el trigo, y deberaca meter mas recursos al trigo en vez del maiz.

d) No , no sera rentable que el granjero plante por $5 por bushel. La solucion optima indica que el granjero debe colocar sus maximos recursos , para maximizar sus ganancias

e) analisis de sencibilidad

Si el precio de el trigo decre a 3$ por bushel , la solucion optima debe ser $x_1 = 133.33, x_2 = 3,666.67, x:3 = 2,500$

La tabla optima debe de ser 


|Z  |  150 | 80 | 60  |  0|
|---|------|----|----|----|
|x1 |  6   |8  | 10  |  5,000|
|x2 |  100 |120 |100  |  60,000|
|x3 |  0   |0 |  0   |  2,500|

 La solución óptima indica que el agricultor necesitaría usar más recursos para cultivar maíz y debería asignar recursos adicionales al maíz, en lugar del trigo.

F. No, no sería rentable para el agricultor plantar maíz a $4 por bushel. La solución óptima indica que el agricultor debe asignar todos los recursos disponibles al trigo para maximizar las ganancias. explique La solución óptima indica que el agricultor debe asignar todos los recursos disponibles al trigo para maximizar la ganancia. 

G)

Análisis de sensibilidad: si el precio del maíz disminuye a $2 por bushel, la solución óptima sería: 

x1 = 200, x2 = 2500, x3 = 3000. 


Cuadro óptimo: 


|Z | 150| 80 |60 | 0|
|---|---|---|---|---|
|x1 | 6 |8 |10 | 5,000|
|x2 | 100| 120| 100 | 60.000|
|x3 | 0| 0| 0 | 3,000|

The optimal solution indicates that the farmer would need to use more resources to grow wheat, and should allocate additional resources to wheat, instead of corn.

Explanation:
If the farmer has contracted to deliver at least the equivalent of 100 acres of wheat, the new optimal solution would be:

|Z  |  150|  80 | 60  |  0|
|---|----|----|----|----|
|x1 |  6  | 8   |10  |  5,000|
|x2 |  100 |120| 100  |  60,000|
|x3 |  0   |0|  | 0   |  3,000|

-----------------------------

6.53
A product is assembled from three parts that can be manufactured on two machines A and B. Neither machine can process different parts at the same time. The number of parts processed by each machine per hour are summarized below:


-------------------------------------

Management seeks a daily schedule (for 8 hrs/day) of the machines so that the number of assemblies is maximized. Currently the company has three machines of type A and five machines of type B.

Solve the problem.

If only one machine can be acquired, which type would you recom-

mend and why?

Management is contemplating the purchase of a type A machine at a

cost of  100,000. Suppose that the life of the machine is 10 years and that each year is equivalent to 2000 working hours. Would you recommend the purchase if the unit profit from each assembly is  1 ? Why or why not?

Following is the calculation of total parts produce by Machine A per day

Part 1 = 10 x 8 = 80 parts

Part 2 = 16 x 8 = 128 parts

Management has 3 A type machine

Part 1 = 3 x 80 = 240 parts

Part 2 = 3 x 128 = 384 parts

Following is the calculation of total parts produce by Machine B per day

Part 1 = 8 x 8 = 64 parts

Part 2 = 12 x 8 = 96 parts

Part 3 = 25 x 8 = 200 parts

Management has 5 B type machine

Part 1 = 5 x 64 = 320 parts

Part 2 = 5 x 96 = 480 parts

Part 3 = 5 x 200 = 1000 parts

Comparing Machine A and Machine B, It is recommended that Machine A should be purchased because output of Part 1 and Part 2 is more compared to machine B. Though Machine A is not producing part 3 but management already have 5 machine B which produces feasible output.

Cost of Machine A = 100,000

Life - 10 years

Total Working hours = 2,000 x 10 = 20,000

Cost of machine A per hour = 100000 / 20000

= $5

Each units has a profit of $10 and $16. Therefore it is recomended to purchase Type A machine.

----------------

 !pip install pulp

In [2]:
# !pip install pulp

Problem Statement

A glass manufacturing company produces two types of glass products A and B.

A= Quantity of type A glass

B= Quantity of type B glass

Objective Function: Profit = 60 * A + 50 * B

Constraints:

Constraint 1: 4 * A + 10 * B <= 100

Constraint 2: 2 * A + 1 * B <= 22

Constraint 3: 3 * A + 3 * B <= 39

A, B >= 0

Initialize Model

In this step, we will import all the classes and functions of pulp module and create a Maxzimization LP problem using LpProblem class.

In [3]:
from pulp import *

# Initialize Class, Define Vars., and Objective
model = LpProblem("Glass_Manufacturing_Industries_Profits",LpMaximize)

Define Decision Variable

In this step, we will define the decision variables. In our problem, we have three variables wood tables, chairs, and bookcases. Let’s create them using LpVariable class. LpVariable will take the following four values:

First, arbitrary name of what this variable represents.
Second is the lower bound on this variable.
Third is the upper bound.
Fourth is essentially the type of data (discrete or continuous). The options for the fourth parameter are LpContinuous or LpInteger.

In [4]:
# Define variables
A = LpVariable('A', lowBound=0)
B = LpVariable('B', lowBound=0)

Define Objective Function


In this step, we will define the maximum objective function by adding it to the LpProblem object.




In [5]:
# Define Objetive Function: Profit on Product A and B
model += 60 * A + 50 * B 

Define the Constraints


In this step, we will add the 4 constraints defined in the problem by adding them to the LpProblem object.

In [6]:
# Constraint 1
model += 4 * A + 10 * B <= 100

# Constraint 2
model += 2 * A + 1 * B <= 22

# Constraint 3
model += 3 * A + 3 * B <= 39

In [7]:
# Solve Model
model.solve()

print("Model Status:{}".format(LpStatus[model.status]))
print("Objective = ", value(model.objective))

for v in model.variables():
    print(v.name,"=", v.varValue)

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/adolfohuerta/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/zb/0l4s85s95rzbg13c3_fm6k540000gn/T/a437275c9e674d69bc8330f141d1f86a-pulp.mps max timeMode elapsed branch printingOptions all solution /var/folders/zb/0l4s85s95rzbg13c3_fm6k540000gn/T/a437275c9e674d69bc8330f141d1f86a-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 8 COLUMNS
At line 17 RHS
At line 21 BOUNDS
At line 22 ENDATA
Problem MODEL has 3 rows, 2 columns and 6 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (0) rows, 2 (0) columns and 6 (0) elements
0  Obj -0 Dual inf 110 (2)
0  Obj -0 Dual inf 110 (2)
2  Obj 740
Optimal - objective value 740
Optimal objective 740 - 2 iterations time 0.002
Option for printingOptions changed from normal to all
Total time (CPU seconds):       0.00   (Wallclock seconds):

In [9]:
import pandas as pd
o = [{'name':name,'shadow price':c.pi,'slack': c.slack} for name, c in model.constraints.items()]

print(pd.DataFrame(o))

  name  shadow price  slack
0  _C1     -0.000000   24.0
1  _C2     10.000000   -0.0
2  _C3     13.333333   -0.0


Puede ver que el valor de sombra para las Restricciones C2 y C3 es 10 y 13,33. significa que si hacemos un cambio de unidad en RHS de las restricciones C2 y C3 afectará la función objetivo por 10 y 13.33. Veamos esto en el ejemplo de la siguiente sección.

El valor de holgura para la restricción C1 es 24, lo que indica el rango de factibilidad para esa restricción. También indica que la restricción es una restricción vinculante. La restricción C2 y C3 tienen holgura = 0. Significa que la restricción es una restricción vinculante. Restricción no vinculante significa que cualquier cambio dentro de este rango no tendrá efecto sobre el valor óptimo de la función objetivo. Podemos verificar esto cambiando el valor de la restricción C1. Veamos el siguiente ejemplo:

In [10]:
# Initialize Class, Define Vars., and Objective
model = LpProblem("Glass_Manufacturing_Industries_Profits",LpMaximize)

# Define variables
A = LpVariable('A', lowBound=0)
B = LpVariable('B', lowBound=0)

# Define Objetive Function: Profit on Product A and B
model += 60 * A + 50 * B 

# Constraint 1
model += 4 * A + 10 * B <= 76 # update its value

# Constraint 2
model += 2 * A + 1 * B <= 22

# Constraint 3
model += 3 * A + 3 * B <= 39

# Solve Model
model.solve()

print("Model Status:{}".format(LpStatus[model.status]))
print("Objective = ", value(model.objective))

for v in model.variables():
    print(v.name,"=", v.varValue)
    
o = [{'Name':name,'Constraint':c,'shadow price':c.pi,'slack': c.slack} for name, c in model.constraints.items()]

print(pd.DataFrame(o))

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/adolfohuerta/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/zb/0l4s85s95rzbg13c3_fm6k540000gn/T/0566f43aed6c4d8d8aa59625e664cb04-pulp.mps max timeMode elapsed branch printingOptions all solution /var/folders/zb/0l4s85s95rzbg13c3_fm6k540000gn/T/0566f43aed6c4d8d8aa59625e664cb04-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 8 COLUMNS
At line 17 RHS
At line 21 BOUNDS
At line 22 ENDATA
Problem MODEL has 3 rows, 2 columns and 6 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (0) rows, 2 (0) columns and 6 (0) elements
0  Obj -0 Dual inf 110 (2)
0  Obj -0 Dual inf 110 (2)
2  Obj 740
Optimal - objective value 740
Optimal objective 740 - 2 iterations time 0.002
Option for printingOptions changed from normal to all
Total time (CPU seconds):       0.00   (Wallclock seconds):

Sensitivity Analysis: Understand the Shadow value

Como ha visto, el valor de sombra para las Restricciones C2 y C3 es 10 y 13,33. significa que si hacemos un cambio de unidad en RHS de las restricciones C2 y C3 afectará la función objetivo por 10 y 13.33. Veamos esto en el siguiente ejemplo. Aquí estamos cambiando el valor RHS de la restricción C2:

In [11]:
# Initialize Class, Define Vars., and Objective
model = LpProblem("Glass_Manufacturing_Industries_Profits",LpMaximize)

# Define variables
A = LpVariable('A', lowBound=0)
B = LpVariable('B', lowBound=0)

# Define Objetive Function: Profit on Product A and B
model += 60 * A + 50 * B 

# Constraint 1
model += 4 * A + 10 * B <= 100

# Constraint 2
model += 2 * A + 1 * B <= 23 # increase by 1

# Constraint 3
model += 3 * A + 3 * B <= 39

# Solve Model
model.solve()

print("Model Status:{}".format(LpStatus[model.status]))
print("Objective = ", value(model.objective))

for v in model.variables():
    print(v.name,"=", v.varValue)
    
o = [{'Name':name,'Constraint':c,'shadow price':c.pi,'slack': c.slack} for name, c in model.constraints.items()]

print(pd.DataFrame(o))

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/adolfohuerta/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/zb/0l4s85s95rzbg13c3_fm6k540000gn/T/f1da36bf52744297982d96fd4410f085-pulp.mps max timeMode elapsed branch printingOptions all solution /var/folders/zb/0l4s85s95rzbg13c3_fm6k540000gn/T/f1da36bf52744297982d96fd4410f085-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 8 COLUMNS
At line 17 RHS
At line 21 BOUNDS
At line 22 ENDATA
Problem MODEL has 3 rows, 2 columns and 6 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (0) rows, 2 (0) columns and 6 (0) elements
0  Obj -0 Dual inf 110 (2)
0  Obj -0 Dual inf 110 (2)
2  Obj 750
Optimal - objective value 750
Optimal objective 750 - 2 iterations time 0.002
Option for printingOptions changed from normal to all
Total time (CPU seconds):       0.00   (Wallclock seconds):

Como ha visto, el valor de sombra para las Restricciones C2 y C3 es 10 y 13,33. significa que si hacemos un cambio de unidad en RHS de las restricciones C2 y C3 afectará la función objetivo por 10 y 13.33. Veamos esto en el siguiente ejemplo. Aquí estamos cambiando el valor RHS de la restricción C2:

Resumen

¡Felicitaciones, has llegado al final de este tutorial!

En este artículo, hemos aprendido el análisis de sensibilidad en el modelado de LP, la interpretabilidad del modelo, el valor de sombra y la variable de holgura con los ejemplos de la biblioteca Python PuLp. Hemos resuelto el problema de programación lineal utilizando PuLP y nos hemos centrado en el análisis de sensibilidad con demostración práctica. Por supuesto, este es solo un ejemplo muy básico de análisis de sensibilidad. En próximos artículos, escribiremos más sobre diferentes problemas de optimización con análisis de sensibilidad. Puede revisar los conceptos básicos de los conceptos matemáticos en este artículo y aprender sobre la programación lineal usando PuLP en este artículo. He escrito más artículos sobre diferentes problemas de optimización, como problemas de transbordo, problemas de asignación, problemas de combinación.